nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/RestLookupService.java

```diff
@@ -0,0 +1,390 @@
+            .collect(Collectors.toMap(
+                e -> e.getKey(),
+                e -> e.getValue().toString()
+            ));
+        try (RecordReader reader = readerFactory.createRecordReader(variables, is, getLogger())) {
+
+            Record record = reader.nextRecord();
+
+            if (recordPath != null) {
+                Optional<FieldValue> fv = recordPath.evaluate(record).getSelectedFields().findFirst();
+                if (fv.isPresent()) {
+                    FieldValue fieldValue = fv.get();
+                    RecordSchema schema = new SimpleRecordSchema(Arrays.asList(fieldValue.getField()));
+                    String[] parts = recordPath.getPath().split("/");
+                    String last = parts[parts.length - 1];
+
+                    Record temp;
+                    Object value = fieldValue.getValue();
+                    if (value instanceof Record) {
+                        temp = (Record) value;
+                    } else if (value instanceof Map) {
+                        temp = new MapRecord(schema, (Map<String, Object>) value);
+                    } else {
+                        temp = new MapRecord(schema, new HashMap<String, Object>() {{
+                            put(last, value);
+                        }});
+                    }
+
+                    record = temp;
+                }
+            }
+
+            return record;
+        } catch (Exception ex) {
+            is.close();
+            throw new RuntimeException(ex);
+        }
+    }
+
+    private Request buildRequest(final String mimeType, final String method, final String body, final String endpoint) {
+        final MediaType mt = MediaType.parse(mimeType);
+        RequestBody requestBody = null;
+        if (body != null) {
+            requestBody = RequestBody.create(mt, body);
+        }
+        Request.Builder request = new Request.Builder()
+                .url(endpoint);
+        switch(method.toLowerCase()) {
+            case "delete":
+                request = body != null ? request.delete(requestBody) : request.delete();
+                break;
+            case "get":
+                request = request.get();
+                break;
+            case "post":
+                request = request.post(requestBody);
+                break;
+            case "put":
+                request = request.put(requestBody);
+                break;
+        }
+
+        if (headers != null) {
+            for (Map.Entry<String, String> header : headers.entrySet()) {
+                request = request.addHeader(header.getKey(), header.getValue());
+            }
+        }
+
+        if (!basicUser.isEmpty() && !isDigest) {
+            String credential = Credentials.basic(basicUser, basicPass);
+            request = request.header("Authorization", credential);
+        }
+
+        return request.build();
+    }
+
+    private String basicUser;
```

Comment:
We should avoid member variables throughout code and try to concentrate them at the top of the class. Additionally, these 3 member variables are not protected so the class is not thread-safe. Need to ensure that all ControllerServices are thread-safe.

Reply:
I think I fixed it.